src/Sparrow.Server/Tensors/Arithmetics.cs

```diff
@@ -0,0 +1,96 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Numerics;
+using System.Runtime.CompilerServices;
+using System.Runtime.Intrinsics.X86;
+using System.Runtime.Intrinsics;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Sparrow.Server.Tensors
+{
+    public static class Arithmetics
+    {
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static Vector256<T> MultiplyAddEstimate<T>(Vector256<T> x, Vector256<T> y, Vector256<T> z)
+            where T : unmanaged
+        {
+#if NET9_0_OR_GREATER
+            if (typeof(T) == typeof(double))
+            {
+                return Vector256.MultiplyAddEstimate(x.AsDouble(), y.AsDouble(), z.AsDouble()).As<double, T>();
+            }
+            else if (typeof(T) == typeof(float))
+            {
+                return Vector256.MultiplyAddEstimate(x.AsSingle(), y.AsSingle(), z.AsSingle()).As<float, T>();
+            }
+#else
+
+            if (Fma.IsSupported)
+            {
+                if (typeof(T) == typeof(float))
+                    return Fma.MultiplyAdd(x.AsSingle(), y.AsSingle(), z.AsSingle()).As<float, T>();
+                if (typeof(T) == typeof(double))
+                    return Fma.MultiplyAdd(x.AsDouble(), y.AsDouble(), z.AsDouble()).As<double, T>();
+            }
+#endif
+            // This version is less accurate numerically.
+            return (x * y) + z;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static Vector512<T> MultiplyAddEstimate<T>(Vector512<T> x, Vector512<T> y, Vector512<T> z)
+            where T : unmanaged
+        {
+#if NET9_0_OR_GREATER
+            if (typeof(T) == typeof(double))
+            {
+                return Vector512.MultiplyAddEstimate(x.AsDouble(), y.AsDouble(), z.AsDouble()).As<double, T>();
+            }
+            else if (typeof(T) == typeof(float))
+            {
+                return Vector512.MultiplyAddEstimate(x.AsSingle(), y.AsSingle(), z.AsSingle()).As<float, T>();
+            }
+#else
+            if (Avx512F.IsSupported)
```

Comment:
Can we add `Avx512F` and similar to `AdvInstructionSet`?

Reply:
Maybe but that would make the compatibility matrix to explode. But on the other hand, the problem we have in those cases is that Avx512 requires the whole set to be implemented (which is kinda strict).